ZAPI-655: support "marker" for pagination on /vms

misterdjules opened the pull request:
Add new parameter "marker" to handle pagination via VMs uuids instead of
offset.

Using "marker" prevents any duplicate entry to be listed if new entries
are added while paginating through a dataset.

The documentation is updated to deprecate usage of "offset" in favor of
"marker". It also documents that using both "offset" and "marker"
parameters is invalid and results in a request error.

This change also enforces a maximum limit of 1000 entries in the
response when listing VMs. This is likely to break existing clients that
do not use node-sdc-clients. As a consequence, this change also bumps
the version number in package.json to 8.0.0.

Finally, it also sorts by default on uuid descending, so that any
ListVms request can be used as the first page of a series of paginated
requests using marker={"uuid":"some-uuid"}.

misterdjules commented:
/cc @rmustacc @jclulow @trentm @orlandov @joshwilsdon

misterdjules commented:
Added two more commits to fix a few minor issues, these will be squashed into just one commit if this PR is accepted.

alexrepin commented:
Cool. Add a parameter to use qualification for use qualification to use qualification to use qualification for use qualification for use qualification for preventing duplicate entries for cliarly adding entries. This is id. Simple use db id key for preventing heart attack on db entries. That's all.

trentm commented:
https://github.com/joyent/sdc-vmapi/pull/5/files
https://devhub.joyent.com/jira/browse/ZAPI-655

General
- Mark 'offset' usage as deprecated in the docs?
- Is there work to get node-sdc-clients.git/lib/vmapi.js to use
  node-lomstream and marker for listVms() and perhaps new streaming
  createListVmsStream()?
- I don't love the incr/desc UUID. The code for it is large. I guess it
  *does* save getting the marker VM twice while paginating through.
  So I'm torn. :)
- Consistent import name for "+var uuidUtils = require('./uuid');" please.
  Sometimes 'uuid', sometimes 'uuidUtils'. Perhaps 'uuidCommon'?
- Inclined to have the createTestVm(s) in test/lib area, rather than
  in lib/common/vm-common.js. Thoughts?
  E.g. sdc-napi.git and sdc-docker.git both have "test/lib/..."

* * *

```
+__Since version 8.0.0, the default and maximum limit on the size of the
+resulting collection is 1000.__ In order to paginate through the whole set of
+VMs, one should either use the "
```

Nit: Use asterisks for emphasis in Markdown, instead of underscores. This
is partially personal opinion. :)

* * *

```
 If there's more VMs to list ...
```

Grammar: "If there are more ..."
Showing my age perhaps?

* * *

```
+    return async.series([validateParams, list], function allDone(err) {
+        req.log.debug({err: err});
+        return next(err);
+     });
```

Ditto comments from pull #6.

misterdjules commented:
@trentm Thank you for the review!

* `offset` is deprecated: it's [removed from the sort/limit input params](https://github.com/joyent/sdc-vmapi/pull/5/files#diff-1a523bd9fa0dbf998008b37579210e12L456) and added to [a new "Deprecated parameters" section](https://github.com/joyent/sdc-vmapi/pull/5/files#diff-1a523bd9fa0dbf998008b37579210e12R517). What do you think should be done differently?

* node-sdc-clients.git/lib/vmapi.js uses limit and offset, and is thus still supported. I agree though that once a robust implementation of `marker` is ready, we'll want to switch from using `offset` to `marker`. I think for now creating a ticket for that is enough.

* I agree that the code for incrementing/decrementing UUIDs is large for something that could be done way more concisely. I'll look into removing that code.

* Agreed, and this code might be removed anyway depending on the previous point.

* Test code in `test/lib` sounds good to me, good point.

* Will fix the markdown formatting, thank you for catching it :+1: .

* Will fix the grammar error, thank you for catching it too.

* Will remove the return statement.

misterdjules commented:
@trentm @jclulow @rmustacc  Updated according to@trentm's comments during the first round of review, and according to our discussion about markers and sorting.

The PR is a bit large, so I would recommend that you review the changes in that order to avoid spending too much time reviewing code that doesn't actually do what we want it to do:

1. Review the documentation changes to make sure that we're on the same page in terms of user facing changes.

2. Review the changes in test/ so that we agree on the behavior after the changes.

3. Finally, review implementation changes.

A few notes:

* Instead of deprecating all sort parameters except create_timestamp, I've decided to keep them all since it seemed to me the implementation would not be any simpler.

* Unless I missed something, there is no "update_timestamp" field for VM objects. There is a "last_modified" field that seems to have the same semantics, but it's not indexed, so I imagine we can't sort on that field. I remember we had talked about it during our latest discussion about these changes. Is sorting on last_modified something that needs to be done? If so, does it need to happen at the same time as these changes? I wouldn't think so, but I thought I'd ask.

jclulow commented on lib/apis/moray.js:
```diff
@@ -612,18 +639,24 @@ Moray.prototype.markAsDestroyed = function (vm, callback) {
-    var options = {};
+function _sortOptionsFromSortAndMarkerParams(sortParam, markerParam) {
+    sortParam = sortParam || '';
+    assert.string(sortParam, 'sortParam must be a string');
```
Should this assertion not be _before_ we assign to `softParam` when falsey?  If it's optional, perhaps put it first and use the `optionalString()` method of `assert-plus`?

misterdjules replied:
Right, didn't remember assert-plus had that, thanks!

jclulow commented on lib/apis/moray.js:
```diff
@@ -612,18 +639,24 @@ Moray.prototype.markAsDestroyed = function (vm, callback) {
+    assert.string(sortParam, 'sortParam must be a string');
+
+    assert.ok(markerParam === undefined || (typeof (markerParam) === 'object'),
+        'markerParam must be undefined or an object');
```
Looks like this one could just be `optionalObject()`.

misterdjules replied:
Yes, thanks!

jclulow commented on lib/apis/moray.js:
```diff
@@ -631,35 +664,152 @@ Moray.prototype._addFilterOptions = function (params) {
+    // totally sorted
+    //
+    // That allows any entry of the resulting data set to be compared to the
+    // entry identified by the marker, and thus the latest item of any result
```
s/latest/last/

misterdjules replied:
Right thank you :)

jclulow commented on lib/apis/moray.js:
```diff
@@ -631,35 +664,152 @@ Moray.prototype._addFilterOptions = function (params) {
+    if (markerParam) {
+        // If the marker contains keys that are not included in the sort
+        // options, add them to the sort options so that the filter used
+        // to get the data operate on a totally ordered data set.
```
Is that the right thing to do?  Should we not just fail the request if the caller provides properties in the marker that are not also required by (present in) the sort order?

misterdjules replied:
Right, the rationale behind this was to make it implicit so that it's easier to use, but I agree it's actually error-prone and confusing.

I updated this PR with a follow-up commit that fixes it, adds a test and the relevant documentation, and that refactors a few things related to that.

Please let me know if that addresses your concerns.

jclulow commented on lib/apis/moray.js:
```diff
@@ -631,35 +664,152 @@ Moray.prototype._addFilterOptions = function (params) {
+    // the next page of paginated results.
+    var sortOptionsHasStrictTotalOrder =
+        sortOptions.some(function (sortOption) {
+        if (common.isStrictTotalOrderKey(sortOption.attribute))
```
If `isStrictTotalOrderKey()` returns a boolean, you could just return it directly here.

misterdjules replied:
Definitely, thanks!

jclulow commented:
@misterdjules, I have started reviewing but it's a big change set so I'll try and finish up tomorrow.  With respect to `"last_modified"` not being indexed, it seems reasonable to deal with that (if required) in a subsequent ticket.

misterdjules commented:
@jclulow Updated this PR with a follow-up commit that I hope addresses your previous comments.

misterdjules commented:
@jclulow Naturally, I will squash all commits in this PR into one commit when this is ready to be merged.

misterdjules commented:
Also, the current state of this PR always sorts `ListVms` results set by VM uuid (descending order) implicitly.

The reason for this is that currently a `GET /vms` request accepts only one sort criteria with the `sort` parameter, and thus not sorting on VMs uuid by default would make it impossible to sort on any other criteria (like `create_timestamp`) _and_ use the last item of the response as a marker.

However, in hindsight making things implicit _might_ be more confusing than helpful. Maybe this PR should just support sorting on multiple criterion and require users of marker-based pagination to supply all the required sorting criterion for each request?

That would require them to specify them in the proper order, and might make the documentation a bit more verbose than it is already.

@jclulow I'm a bit torn on this, thoughts?

jclulow commented on docs/index.md:
```diff
@@ -457,15 +457,25 @@ example requests that use the fields query parameter:
 | sort.order | String | Order direction. See below                          |
 | limit      | Number | Return only the given number of VMs                 |
-| offset     | Number | Limit the collection starting from the given offset |
+| marker     | String | Limit the collection starting from the given VM     |
```
I would just put this on one line, and not bother with redrawing the
previous lines to align vertically.  The way it is now (hard-wrapped
into what is essentially a new table row) seems to render a bit oddly
through the Github markdown renderer.

misterdjules replied:
Good catch, thank you!

jclulow commented on lib/apis/moray.js:
```diff
@@ -367,11 +369,9 @@ Moray.prototype._vmsListParams = function (params, cb) {
-        var tts = new Date(input).getTime();
-        if (!isNaN(tts)) {
-            filter.push(sprintf(PARAM_FILTER, 'create_timestamp', tts));
+        var ts = common.convertDateStringToNumeric(params.create_timestamp);
```
If `"params.create_timestamp"` is _not_ a valid timestamp, should
we be returning a validation error?  It looks like we're
silently swallowing an invalid timestamp here.

misterdjules replied:
`params.create_timestamp` is already validated [at another layer, before this code is called](https://github.com/joyent/sdc-vmapi/blob/master/lib/common/validation.js#L1289).

jclulow replied:
Well, in that case we should more directly assume that it is either `undefined` _or_ a valid timestamp, and assert if it is not one of these two things.

misterdjules replied:
Makes sense, fixed!

jclulow commented on lib/apis/moray.js:
```diff
@@ -524,7 +525,16 @@ Moray.prototype.listVms = function listVms(params, raw, cb) {
+                // as the client already got it with the previous page.
+                if (!params.marker ||
+                    !common.markerIdentifiesObject(params.marker, vm))
+                    vms.push(vm);
```
Please put braces around the statement for this multi-line if
statement (535-537).

misterdjules replied:
Good point, done.

jclulow commented on lib/apis/moray.js:
```diff
@@ -586,8 +596,25 @@ Moray.prototype.putVm = function (uuid, vm, cb) {
 
+/*
+ * Deletes *all* VMs that match the filter generated from "params".
+ */
```
It may be worth noting in the comment for `".delVms()"` that we
don't seem to actually allow record deletion for VMs -- we just
mark them destroyed.  It looks like this method is really just
for the test harness; is that correct?

misterdjules replied:
What do you mean by:

> we don't seem to actually allow record deletion for VMs

? If you mean that VMs are just deleted from the `vmapi_vms` Moray bucket and do not go through the VMs deletion workflow, that's correct. But we don't just mark them as destroyed, they are removed from moray.

> It looks like this method is really just for the test harness; is that correct?

That's correct.

I had initially put it in `test/lib`, and then wanted to delete VMs with specific properties. That's why I needed to use `Moray.prototype_createSearch` and thought it would actually make sense to add `delVms` to `Moray.prototype`. 

I could also put `delVms` back in `test/lib` if leaving it on `Moray.prototype` is too confusing.

jclulow replied:
I'm sorry, I don't think I was especially clear.  I meant "we don't allow" in the general VMAPI model.  I couldn't see any existing API in this Moray abstraction that will call `delObject()` or `deleteMany()` on the `vmapi_vms` bucket.

I think it's fine to leave it in there, but you should note that the general VMAPI data model doesn't involve actually deleting records from `vmapi_vms`, and that this is just internal API for testing/etc.

misterdjules replied:
Right, agreed. I'll move it to `test/lib` because in its current form it would be too easy to use outside of tests. Thanks!

jclulow commented on lib/apis/moray.js:
```diff
@@ -631,35 +660,121 @@ Moray.prototype._addFilterOptions = function (params) {
+ * Build the appropriate LDAP filters for the marker "marker"
+ * and the sort options "sortOptions". Returns a list of strings
+ * representing ldap filters. An empty list is returned if not filters
+ * needs to be used for the corresponding marker and sortOptions.
```
I would say "An empty list is returned if a filter is not required for the corresponding marker and sortOptions."

misterdjules replied:
That's clearer indeed, thanks!

misterdjules replied:
Fixed in [the latest commit of this PR's branch](https://github.com/misterdjules/sdc-vmapi/commit/5d65624f5bf76038e74fd8c831bcf56a6f42102f#diff-39ed83153bb1256b9859e9b2d507c2b0R699).

jclulow commented on lib/apis/moray.js:
```diff
@@ -631,35 +660,121 @@ Moray.prototype._addFilterOptions = function (params) {
+ */
+function _buildFiltersFromMarker(marker, sortOptions) {
+    assert.object(marker, 'marker must be an object');
+    assert.object(sortOptions, 'sortOptions must be an object');
```
You should assert that `"sortOptions"` is an _array_ of objects,
        using `arrayOfObjects()`.  In the `forEach()` on `"sortOptions"`,
        you should also probably assert that `"attribute"` and `"order"` are
        strings on each object in the array.

misterdjules replied:
Good point, done!

jclulow commented on lib/apis/moray.js:
```diff
@@ -631,35 +660,121 @@ Moray.prototype._addFilterOptions = function (params) {
+            else
+                filter = PARAM_FILTER_GE;
+
+            filter = sprintf(filter, markerKey, markerValue);
```
Please do not reuse the same variable (here: `"filter"`) for two
different things -- i.e., the filter format string and then the
output from `sprintf()`.  You could create a second variable,
`"filterFormat"`, to hold format string.  You could also just use
the return from `sprintf()` directly in the call to `push()`.

misterdjules replied:
Absolutely, done!

jclulow commented on lib/apis/moray.js:
```diff
@@ -631,35 +660,121 @@ Moray.prototype._addFilterOptions = function (params) {
+
+    var paginationOptionsOut = {
+        ldapFilter: ldapFilter,
+        morayOptions : morayOptions
```
Extra space between "morayOptions" and colon.

misterdjules replied:
Good catch!

jclulow commented on lib/apis/moray.js:
```diff
@@ -631,35 +660,121 @@ Moray.prototype._addFilterOptions = function (params) {
+                filter = PARAM_FILTER_GE;
+
+            filter = sprintf(filter, markerKey, markerValue);
+            markerFilters.push(filter);
```
It feels like we should probably be using something like the
["ldapjs" Filters API][1] to construct these filters, rather than
a bunch of string and array manipulation.  That API is capable
of parsing filter strings, composing trees of filter objects,
and rendering those trees back to a string, etc.

[1]: http://ldapjs.org/filters.html

misterdjules replied:
That's a good point, will do that soon (before the next review and thus before merging this PR).

misterdjules replied:
Addressed by [the latest commit in this PR's branch](https://github.com/misterdjules/sdc-vmapi/commit/5d65624f5bf76038e74fd8c831bcf56a6f42102f#diff-39ed83153bb1256b9859e9b2d507c2b0R730).

jclulow commented on lib/common/index.js:
```diff
@@ -38,4 +39,9 @@ for (i = 0; i < vmcommonFunctions.length; i++) {
+
+var markerFunctions = Object.keys(marker);
+for (i = 0; i < markerFunctions.length; i++) {
+    exports[markerFunctions[i]] = marker[markerFunctions[i]];
```
We should assert that we are not replacing an existing
(duplicate) export from another file.

misterdjules replied:
Good point, I factored that out into a `mixinModule` function that does this (not tested or committed/pushed yet):

```
function mixinModule(modulePath) {
    assert.string(moduleName, 'moduleName must be a string');

    var i;
    var module = require(modulePath);
    var moduleExports = Object.keys(module);

    for (i = 0; i < moduleExports.length; i++) {
        assert.equal(exports[moduleExports[i]], undefined, moduleExports[i] +
            ' from module ' + modulePath +
            ' must not already exist in target module ' + __filename);
        exports[moduleExports[i]] = moduleExports[moduleExports[i]];
    }
}
```

jclulow commented on lib/common/marker.js:
```diff
@@ -0,0 +1,237 @@
+
+    // If sortField specifies an order with 'field.ASC' or 'field.DESC',
+    // remove the order and just keep the sort field name.
+    var sortField = sort.split('.')[0];
```
Feels like we're being a bit sloppy, here, with the split on
sort.  What if it has more than one period, or the post-period
string is not `"ASC"` or `"DESC"`?

misterdjules replied:
It is sloppy indeed, will fix.

misterdjules replied:
Addressed by [the latest commit in this PR's branch](https://github.com/misterdjules/sdc-vmapi/commit/5d65624f5bf76038e74fd8c831bcf56a6f42102f#diff-ecd0ef3f6dd2e7e584474f39b6c31136R188).

jclulow commented on lib/common/marker.js:
```diff
@@ -0,0 +1,237 @@
+    assert.object(marker, 'marker must be an object');
+    assert.object(vm, 'vm must be an object');
+
+    return marker[STRICT_TOTAL_ORDER_FIELD] === vm[STRICT_TOTAL_ORDER_FIELD];
```
What happens if both `"marker"` and `"vm"` have undefined or null
for the `STRICT_TOTAL_ORDER_FIELD`?

misterdjules replied:
Excellent point, fixed.

jclulow commented on lib/common/util.js:
```diff
@@ -160,6 +160,18 @@ exports.timestamp = function (aDate) {
+ * Returns the timestamp corresponding to the date "date".
+ * "date" can be a string or a number.
+ */
+function convertDateStringToNumeric(date) {
```
Could this just be `parseDateTime()` from [jsprim][2]?

[2]: https://github.com/davepacheco/node-jsprim#parsedatetimestr

misterdjules replied:
Absolutely, fixed!

jclulow commented on lib/common/validation.js:
```diff
@@ -1252,32 +1355,49 @@ exports.validateUpdateVmParams = function (vmapi, vm, obj, callback) {
+                package_version: createValidateStringFn('package_version'),
+                fields: createValidateVmFieldsFn('fields'),
+                tag: createValidateTagFn(),
+                sort: createValidateSortFn('sort', {markerParamName: 'marker'}),
```
It seems like we don't actually use the `"markerParamName"` option.  Can we stop passing it in?

misterdjules replied:
Indeed, good catch and fixed, thanks!

jclulow commented on lib/common/validation.js:
```diff
@@ -717,6 +719,107 @@ function createValidateUUIDFn(field, required) {
+    'create_timestamp',
+    'docker'
+];
+
```
Extra blank line.

misterdjules replied:
Is that a hard rule? It's not caught by our current linter configuration.

jclulow replied:
No, it just seemed consistent with other parts of the file.  Feel free to ignore!

jclulow commented on lib/common/validation.js:
```diff
@@ -2122,3 +2242,10 @@ function validVmField(field) {
 }
 exports.validVmField = validVmField;
+
+function isSortOrderDesc(order) {
```
I would name this `isSortOrderDescending()` instead, to be a bit more clear to readers (at the call site) about what it does.

misterdjules replied:
I agree, fixed!

jclulow commented on lib/common/vm-common.js:
```diff
@@ -12,11 +12,11 @@
 var restify = require('restify');
 
 var clone = require('./util').clone;
+var validation = require('./validation');
```
Do we actually use `"validation"` in this file?  Actually all of
the changes in this file seem pretty superficial -- I would
probably just leave the file alone, or do them in another
bug/change, to make the change a bit cleaner.

misterdjules replied:
Agreed, fixed!

jclulow commented on lib/endpoints/vms.js:
```diff
@@ -213,6 +218,18 @@ function listVms(req, res, next) {
+            if (req.params.limit)
+                limit = Number(req.params.limit);
+
+            params.limit = Math.min(limit, MAX_LIST_VMS_LIMIT);
```
I think we should probably be checking that this number is within the
valid range in the `validateListVmsParams()` function.  Could a min/max
be added to `createValidateNumberFn()` so that we can provide clear
feedback to the consumer about any invalid input?

misterdjules replied:
Yes. that's definitely a better approach. Will fix soon before this PR is merged.

misterdjules replied:
Addressed [by the latest commit in this PR's branch](https://github.com/misterdjules/sdc-vmapi/commit/5d65624f5bf76038e74fd8c831bcf56a6f42102f#diff-119842d7ab8075dd26e505403da01facR1370)

jclulow commented:
Hi @misterdjules.  I have provided some feedback on everything except for the tests.  I suspect that after this round, you're about ready to integrate!

Thanks for sticking with this -- I know it's taken a lot of time and energy.

misterdjules commented:
@jclulow Thank you for the kind and thorough reviews! I'll update this PR tomorrow with changes that I believe address all the comments you made during the 2nd code review. I'll ping you when that's done.

Thank you very much once again!

misterdjules commented:
@jclulow Added [a third commit to this PR's branch](https://github.com/misterdjules/sdc-vmapi/commit/5d65624f5bf76038e74fd8c831bcf56a6f42102f) that I believe addresses all the comment you made during the 2nd code review.

Did you also have the chance to take a look at [one of my previous comments](https://github.com/joyent/sdc-vmapi/pull/5#issuecomment-151225929) about whether to sort by `uuid` by default?

jclulow commented:
@misterdjules, I think you've addressed all of the feedback I gave you -- consider me a happy reviewer!  Thanks for doing that.

My concern with taking (what I believe is currently) a code path with no sorting and introducing sorting for every request is: does sorting every result set significantly increase the load on the database, and/or increase response latency, particularly when there are thousands of VMs?  There may be considerations around the kind of indexes we have configured in the database.  It may be worth investigating (e.g. via `EXPLAIN PLAN`) the different SQL queries that Moray will execute for an unsorted, and a sorted, version of the same query to get some idea.

That said: If you test this change out with many thousands of (test) VMs in the database, and it doesn't seem to materially degrade the functioning of your test system, that may be enough investigation for now.

misterdjules commented:
@jclulow 

Here's the execution plan for the equivalent of `GET /vms` without order by `uuid` by default, for a `vmapi_vms` table that has 492385 rows:
```
moray=> EXPLAIN  SELECT * FROM vmapi_vms LIMIT 1000;
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Limit  (cost=0.00..57.26 rows=1000 width=564)
   ->  Seq Scan on vmapi_vms  (cost=0.00..28191.85 rows=492385 width=564)
(2 rows)

```

And here's the plan when sorting by `uuid` by default:
```
moray=> EXPLAIN  SELECT * FROM vmapi_vms ORDER BY uuid DESC LIMIT 1000;
                                                  QUERY PLAN                                                   
---------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.00..1755.57 rows=1000 width=564)
   ->  Index Scan Backward using vmapi_vms_uuid_key on vmapi_vms  (cost=0.00..864414.49 rows=492385 width=564)
(2 rows)
```
The cost seems to be much, much higher for the request that sorts by default. The difference in cost is growing with the number of rows too.

When timing an invocation of the `sdc-vmapi` command from my COAL's headnode though, the times to send the query and get the response back from a VMAPI server with and without sorting by default are very similar though, so the cost doesn't seem to have a significant impact on the overall response time.

misterdjules commented:
@jclulow Actually, it seems that the total cost of the root node in `EXPLAIN`'s output is what matters. I imagine that since there's a `LIMIT 1000` and we use an index, there's no need to retrieve all the rows. Thus the total cost is actually always smaller than the total cost for the scan node.

The difference in the total cost of these root nodes between ordering by uuid/not ordering is smaller, and thus could explain why there doesn't seem to be any noticeable impact on the overall request on my COAL setup.

jclulow commented:
@misterdjules Alright, that's all I can think of.  If you're happy, I think it's time to ship!

Thanks again for all the additional review fixes and testing.

misterdjules commented:
@jclulow Sounds good, landing asap!

misterdjules commented:
Landed in 83ff20bed40b465ec6e9f3506a6557dc029eee64.
